source/vir/src/split_expression.rs

```diff
@@ -0,0 +1,546 @@
+                    BndX::Quant(_, binders, _) => binders.iter().map(|b| b.name.clone()).collect(),
+                    BndX::Choose(binders, _, _) => binders.iter().map(|b| b.name.clone()).collect(),
+                    BndX::Lambda(binders) => binders.iter().map(|b| b.name.clone()).collect(),
+                };
+                !binded_names.iter().any(|bname| **bname == **name)
+            });
+            let arg_map = arg_map;
+            let e1_replaced = subsitute_argument(e1, &arg_map)?;
+            let bnd_replaced: Bnd = match &bnd.x {
+                BndX::Let(binders) => {
+                    let mut new_binders: Vec<Binder<Exp>> = vec![];
+                    for old_b in &**binders {
+                        let new_b = BinderX {
+                            name: old_b.name.clone(),
+                            a: subsitute_argument(&old_b.a, &arg_map)?,
+                        };
+                        new_binders.push(Arc::new(new_b));
+                    }
+                    Spanned::new(bnd.span.clone(), BndX::Let(Arc::new(new_binders)))
+                }
+                BndX::Quant(quant, bndrs, _trigs) => Spanned::new(
+                    bnd.span.clone(),
+                    BndX::Quant(quant.clone(), bndrs.clone(), Arc::new(vec![])),
+                ),
+                _ => {
+                    return Err((
+                        exp.span.clone(),
+                        format!("Unsupported binder during subsitution"),
+                    ));
+                }
+            };
+            let new_exp = ExpX::Bind(bnd_replaced, e1_replaced);
+            SpannedTyped::new(&exp.span, &exp.typ, new_exp)
+        }
+        ExpX::Const(_) => exp.clone(),
+        _ => {
+            return Err((exp.span.clone(), format!("Unsupported expression during subsitution")));
+        }
+    };
+    Ok(result)
+}
+
+fn is_bool_type(t: &Typ) -> bool {
+    match &**t {
+        TypX::Bool => true,
+        TypX::Boxed(b) => is_bool_type(b),
+        _ => false,
+    }
+}
+
+fn is_same_type(t1: &Typ, t2: &Typ) -> bool {
+    match (&**t1, &**t2) {
+        (TypX::Bool, TypX::Bool)
+        | (TypX::Int(_), TypX::Int(_))
+        | (TypX::Tuple(_), TypX::Tuple(_))
+        | (TypX::Lambda(_, _), TypX::Lambda(_, _))
+        | (TypX::Datatype(_, _), TypX::Datatype(_, _))
+        | (TypX::TypParam(_), TypX::TypParam(_))
+        | (TypX::TypeId, TypX::TypeId)
+        | (TypX::Air(_), TypX::Air(_)) => true,
+        (TypX::Boxed(b1), TypX::Boxed(b2)) => is_same_type(b1, b2),
+        _ => false,
+    }
+}
+
+pub(crate) fn tr_inline_expression(
+    body_exp: &Exp,
+    params: &Params,
+    exps: &Exps,
+) -> Result<Exp, (Span, String)> {
+    match *body_exp.typ {
+        TypX::Bool => (),
+        _ => {
+            return Err((
+                body_exp.span.clone(),
+                "Skip inlining for non-boolean expreesion".to_string(),
```

Comment:
typo, `expreesion` -> `expression`

Reply:
Also, I'm not sure what the context of this error is or why it would trigger, but it doesn't really sound like an error message. If it's "skipping" something, that makes it sound like a warning of some sort.

Reply:
I added this after getting an error from an attempt to inline a function with "empty body". I think it consisted of `unimplemented!`, and `verifier(external_body)` attribute.  

All `Err` in `subsitute_argument` and `tr_inline_expression` are now downgraded to "Note",  specifying the reason why the function inlining failed.  I now realize I should add more documentation here. Thanks!

Reply:
Can you make a specific message for `external_body`, then? (Similar to the ones about opaque functions)

Reply:
That sounds great -- thanks!